src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java

```diff
@@ -0,0 +1,106 @@
+package org.tools4j.hoverraft.state;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class InMemoryCommandLog implements CommandLog {
+    private final List<CommandLogEntry> commandLogEntries = new ArrayList<>();
+    private final AtomicInteger readIndex = new AtomicInteger(0);
+
+
+    @Override
+    public long size() {
+        return commandLogEntries.size();
+    }
+
+    @Override
+    public long readIndex() {
+        return readIndex.get();
+    }
+
+    @Override
+    public synchronized void readIndex(final long index) {
+        if (index < commandLogEntries.size()) {
+            readIndex.set((int)index);
+        } else {
+            throw new IllegalArgumentException("Invalid read index " + index + " for log entry with size " + commandLogEntries.size());
+        }
+    }
+
+    @Override
+    public synchronized CommandLogEntry read() {
+        if (readIndex.get() < commandLogEntries.size()) {
+            return commandLogEntries.get(readIndex.getAndIncrement());
+        }
+        throw new IllegalArgumentException("Read index " + readIndex + " has reached end of message log with size " + commandLogEntries.size());
```

Comment:
there are no arguments. this is either IllegalStateEx or it should never be possible to get into this situation.

It should actually only occur if index=0 and log is empty. I would throw an IllegalStateException in this case